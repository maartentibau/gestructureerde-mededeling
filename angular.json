{
  "$schema": "./node_modules/@angular/cli/lib/config/schema.json",
  "version": 1,
  "newProjectRoot": "projects",
  "projects": {
    "gestructureerde-mededeling": {
      "projectType": "application",
      "schematics": {
        "@schematics/angular:component": {
          "style": "scss"
        }
      },
      "root": "",
      "sourceRoot": "src",
      "prefix": "app",
      "architect": {
        "build": {
          "builder": "@angular-devkit/build-angular:application",
          "options": {
            "allowedCommonJsDependencies": ["clipboard"],
            "outputPath": "dist/gestructureerde-mededeling",
            "index": "src/index.html",
            "browser": "src/main.ts",
            "tsConfig": "tsconfig.app.json",
            "inlineStyleLanguage": "scss",
            "assets": ["./src/favicon.ico", "./src/assets", "./src/_redirects"],
            "styles": [
              "normalize.css/normalize.css",
              "@angular/material/prebuilt-themes/azure-blue.css",
              "./src/styles.scss"
            ],
            "scripts": []
          },
          "configurations": {
            "production": {
              "budgets": [
                {
                  "type": "initial",
                  "maximumWarning": "500kB",
                  "maximumError": "1MB"
                },
                {
                  "type": "anyComponentStyle",
                  "maximumWarning": "2kB",
                  "maximumError": "4kB"
                }
              ],
              "outputHashing": "all"
            },
            "development": {
              "optimization": false,
              "extractLicenses": false,
              "sourceMap": true
            }
          },
          "defaultConfiguration": "production"
        },
        "serve": {
          "builder": "@angular-devkit/build-angular:dev-server",
          "configurations": {
            "production": {
              "buildTarget": "gestructureerde-mededeling:build:production"
            },
            "development": {
              "buildTarget": "gestructureerde-mededeling:build:development"
            }
          },
          "defaultConfiguration": "development"
        },
        "extract-i18n": {
          "builder": "@angular-devkit/build-angular:extract-i18n"
        },
        "test": {
          "builder": "@angular-devkit/build-angular:jest",
          "options": {
            "tsConfig": "tsconfig.spec.json"
          }
        }
      }
    }
  }
}
