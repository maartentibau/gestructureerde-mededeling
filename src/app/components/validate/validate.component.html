<ogm-number [ogm]="ogm()" [isValid]="isValid()"></ogm-number>

<ogm-input
  (ogmChange)="handleOgmChange($event)"
  [validate]="true"
  placeholderMessage="vul de 12 cijfers in van uw gestructureerde mededeling"
></ogm-input>
