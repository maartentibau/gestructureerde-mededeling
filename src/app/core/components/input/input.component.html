<div class="ogm-form">
  <input
    type="text"
    [attr.maxlength]="validate ? 12 : 10"
    [formControl]="ogmInput"
    [attr.placeholder]="placeholderMessage"
  />
</div>
