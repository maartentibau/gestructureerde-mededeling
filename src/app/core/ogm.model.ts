export interface OgmData {
  number: string;
  numberFormat: string;
}
