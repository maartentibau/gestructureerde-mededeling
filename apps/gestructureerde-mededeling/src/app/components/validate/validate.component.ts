import { ChangeDetectionStrategy, Component } from '@angular/core';
import { BehaviorSubject, Subject } from 'rxjs';
import { OgmService } from '../../core/services/ogm.service';
import { OgmInputChange } from '../../core/components/input/input.component';

@Component({
  selector: 'ogm-validate',
  templateUrl: './validate.component.html',
  styleUrls: ['./validate.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ValidateComponent {
  readonly ogm$: BehaviorSubject<string>;
  readonly isValid$: Subject<boolean>;

  constructor(private ogmService: OgmService) {
    this.ogm$ = new BehaviorSubject<string>(this.ogmService.init());
    this.isValid$ = new Subject<boolean>();
  }

  ogmInputChangeHandler({ ogm, isValid }: OgmInputChange) {
    this.ogm$.next(ogm);
    this.isValid$.next(isValid);
  }
}
