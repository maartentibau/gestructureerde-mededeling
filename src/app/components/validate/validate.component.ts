import { AsyncPipe } from '@angular/common';
import { ChangeDetectionStrategy, Component, inject, signal } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { Title } from '@angular/platform-browser';
import { InputComponent, Ogm } from '../../core/components/input/input.component';
import { NumberComponent } from '../../core/components/number/number.component';
import { DEFAULT_TITLE } from '../../core/core.constants';
import { OgmService } from '../../core/services/ogm.service';

@Component({
  standalone: true,
  imports: [NumberComponent, InputComponent, AsyncPipe, FormsModule],
  selector: 'ogm-validate',
  templateUrl: './validate.component.html',
  styleUrls: ['./validate.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ValidateComponent {
  #ogmService: OgmService = inject(OgmService);
  #title: Title = inject(Title);

  readonly ogm = signal<string | null>(this.#ogmService.init());
  readonly isValid = signal<boolean | null>(null);

  constructor() {
    this.#title.setTitle(`${DEFAULT_TITLE} - Controleer een gestructureerde mededeling`);
  }

  handleOgmChange({ ogm, isValid }: Ogm) {
    this.ogm.set(ogm);
    this.isValid.set(isValid);
  }
}
