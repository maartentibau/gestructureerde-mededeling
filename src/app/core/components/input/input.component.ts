import { Component, EventEmitter, inject, Input, OnDestroy, OnInit, Output } from '@angular/core';
import { ReactiveFormsModule, UntypedFormControl } from '@angular/forms';
import { Observable, Subject } from 'rxjs';
import { filter, map, takeUntil, tap } from 'rxjs/operators';
import { OgmService } from '../../services/ogm.service';

export interface OgmInputChange {
  ogm: string;
  isValid: boolean | null;
}

@Component({
  standalone: true,
  imports: [ReactiveFormsModule],
  selector: 'ogm-input',
  templateUrl: './input.component.html',
  styleUrls: ['./input.component.scss'],
})
export class InputComponent implements OnInit, OnDestroy {
  #ogmService: OgmService = inject(OgmService);

  @Input() validate!: boolean;
  @Input() placeholderMessage!: string;

  @Output() readonly ogmInputChange: EventEmitter<OgmInputChange>;

  readonly ogmInput: UntypedFormControl;
  readonly destroy$: Subject<void>;

  ogm$!: Observable<string>;

  constructor() {
    this.ogmInputChange = new EventEmitter<OgmInputChange>();

    this.ogmInput = new UntypedFormControl(null);
    this.destroy$ = new Subject<void>();
  }

  ngOnInit() {
    const ogmInitValue: string = this.#ogmService.init();
    const hasSpaces: RegExp = new RegExp('[\\s]');

    const omgInputChanges$ = this.ogmInput.valueChanges.pipe(
      filter((value) => {
        if (isNaN(Number(value)) || hasSpaces.test(value)) {
          const cleanValue = this.#ogmService.clean(value);
          this.ogmInput.setValue(cleanValue, { emitEvent: false });
          return false;
        }

        return true;
      })
    );

    if (this.validate) {
      this.ogm$ = omgInputChanges$.pipe(
        map((value) => (value ? this.#ogmService.format(value.toString().padEnd(12, ' ')) : ogmInitValue))
      );
    } else {
      this.ogm$ = omgInputChanges$.pipe(
        map((value) => (value ? this.#ogmService.generate(value, true) : ogmInitValue))
      );
    }

    this.ogm$
      .pipe(
        tap((ogm: string) => {
          let isValid = null;

          if (this.validate) {
            const ogmNumber = this.#ogmService.clean(ogm);
            isValid = ogmNumber.length === 12 ? this.#ogmService.validate(ogm) : null;
          }

          this.ogmInputChange.emit({ ogm, isValid });
        }),
        takeUntil(this.destroy$)
      )
      .subscribe();
  }

  ngOnDestroy() {
    this.destroy$.next();
  }
}
