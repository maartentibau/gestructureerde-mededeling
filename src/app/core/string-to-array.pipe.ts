import { Pipe, PipeTransform } from '@angular/core';

@Pipe({
  standalone: true,
  name: 'stringToArray',
})
export class StringToArrayPipe implements PipeTransform {
  transform(value: number | string | null | undefined): string[] {
    if ((value && typeof value === 'string') || typeof value === 'number') {
      return Array.from(String(value));
    }

    return [];
  }
}
