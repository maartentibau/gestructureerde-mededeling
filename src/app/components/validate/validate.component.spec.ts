import { AsyncPipe, JsonPipe } from '@angular/common';
import { Component, input } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { MatSnackBar } from '@angular/material/snack-bar';
import { Title } from '@angular/platform-browser';
import { render } from '@testing-library/angular';
import { Ogm } from '../../core/components/input/input.component';
import { DEFAULT_TITLE } from '../../core/core.constants';

import { ValidateComponent } from './validate.component';

@Component({
    imports: [JsonPipe],
    selector: 'ogm-number',
    template: `
    <div>{{ ogm | json }}</div>
    <div>{{ isValid | json }}</div>
  `
})
class MockNumberComponent {
  ogm = input<string | null>(null);
  isValid = input<boolean | null>(null);
}

@Component({
    imports: [JsonPipe],
    selector: 'ogm-input',
    template: `
    <div>{{ validate | json }}</div>
    <div>{{ placeholderMessage | json }}</div>
  `
})
class MockInputComponent {
  validate = input<boolean>(false);
  placeholderMessage = input<string>('');
}

describe('ValidateComponent', () => {
  let component: ValidateComponent;
  let fixture: ComponentFixture<ValidateComponent>;
  let titleService: Title;

  beforeEach(async () => {
    ({ component, fixture, titleService } = await setup());
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  describe('constructor', () => {
    it('set the correct HTML title', () => {
      // prepare
      const title: string = `${DEFAULT_TITLE} - Controleer een gestructureerde mededeling`;
      jest.spyOn(titleService, 'setTitle');

      // act
      TestBed.createComponent(ValidateComponent);

      // check
      expect(titleService.setTitle).toHaveBeenCalledWith(title);
    });
  });

  describe('rendering component', () => {
    it('should match snapshot', () => {
      // act
      fixture.detectChanges();

      // check
      expect(fixture).toMatchSnapshot();
    });
  });

  describe('ogmInputChangeHandler', () => {
    it('should call next on ogm$ and isValid$ stream', () => {
      // prepare
      jest.spyOn(component.ogm, 'set');
      jest.spyOn(component.isValid, 'set');

      const ogmInputChange: Ogm = { ogm: '12345', isValid: false };

      // act
      component.handleOgmChange(ogmInputChange);

      // check
      expect(component.ogm.set).toHaveBeenCalledWith('12345');
      expect(component.isValid.set).toHaveBeenCalledWith(false);
    });
  });
});

const setup = async () => {
  const renderResult = await render(ValidateComponent, {
    componentImports: [MockNumberComponent, MockInputComponent, AsyncPipe],
    providers: [{ provide: MatSnackBar, useValue: { open: jest.fn() } }],
  });

  const titleService = TestBed.inject(Title);

  return {
    ...renderResult,
    component: renderResult.fixture.componentInstance,
    titleService,
  };
};
