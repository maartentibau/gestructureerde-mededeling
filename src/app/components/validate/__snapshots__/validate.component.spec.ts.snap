// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ValidateComponent rendering component should match snapshot 1`] = `
<ogm-validate
  isValid={[Function Function]}
  ogm={[Function Function]}
>
  <ogm-number>
    <div>
    </div>
    <div>
    </div>
  </ogm-number><ogm-input
    placeholdermessage="vul de 12 cijfers in van uw gestructureerde mededeling"
  >
    <div>
    </div>
    <div>
    </div>
  </ogm-input>
</ogm-validate>
`;
